nixos/modules/services/search/opensearch.nix

```diff
@@ -0,0 +1,221 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.opensearch;
+
+  opensearchConfig = ''
+    network.host: ${cfg.listenAddress}
+    cluster.name: ${cfg.cluster_name}
+    ${lib.optionalString cfg.single_node "discovery.type: single-node"}
+    ${lib.optionalString cfg.single_node "gateway.auto_import_dangling_indices: true"}
+
+    http.port: ${toString cfg.port}
+    transport.port: ${toString cfg.tcp_port}
+
+    ${cfg.extraConf}
+  '';
+
+  configDir = cfg.dataDir + "/config";
+
+  opensearchYml = pkgs.writeTextFile {
+    name = "opensearch.yml";
+    text = opensearchConfig;
+  };
+
+  loggingConfigFilename = "log4j2.properties";
+  loggingConfigFile = pkgs.writeTextFile {
+    name = loggingConfigFilename;
+    text = cfg.logging;
+  };
+
+in
+{
+
+  ###### interface
+
+  options.services.opensearch = {
+    enable = mkEnableOption (lib.mdDoc "OpenSearch, a distributed search and analytics engine.");
+
+    package = mkOption {
+      description = lib.mdDoc "OpenSearch package to use.";
+      default = pkgs.opensearch;
+      defaultText = literalExpression "pkgs.opensearch";
+      type = types.package;
+    };
+
+    listenAddress = mkOption {
+      description = lib.mdDoc "OpenSearch listen address.";
+      default = "127.0.0.1";
+      type = types.str;
+    };
+
+    port = mkOption {
+      description = lib.mdDoc "OpenSearch port to listen for HTTP traffic.";
+      default = 9200;
+      type = types.int;
+    };
+
+    tcp_port = mkOption {
+      description = lib.mdDoc "OpenSearch port for the node to node communication.";
+      default = 9300;
+      type = types.int;
+    };
+
+    cluster_name = mkOption {
+      description = lib.mdDoc "OpenSearch name that identifies your cluster for auto-discovery.";
+      default = "opensearch";
+      type = types.str;
+    };
+
+    single_node = mkOption {
+      description = lib.mdDoc "Start a single-node cluster";
+      default = true;
+      type = types.bool;
+    };
+
+    extraConf = mkOption {
```

Comment:
Maybe just replace this with a `settings` option? `freeform` options? See https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md for details.